Source/WebKit/UIProcess/mac/DisplayLink.h

```diff
@@ -27,70 +27,71 @@
-    
-    void addObserver(IPC::Connection::UniqueID, DisplayLinkObserverID, WebCore::FramesPerSecond);
-    void removeObserver(IPC::Connection::UniqueID, DisplayLinkObserverID);
-    void removeObservers(IPC::Connection::UniqueID);
-
-    void incrementFullSpeedRequestClientCount(IPC::Connection::UniqueID);
-    void decrementFullSpeedRequestClientCount(IPC::Connection::UniqueID);
+    class Client {
+    friend class DisplayLink;
+    public:
+        virtual ~Client() = default;
+        
+    private:
+        virtual void displayLinkFired(WebCore::PlatformDisplayID, WebCore::DisplayUpdate, bool wantsFullSpeedUpdates, bool anyObserverWantsCallback) = 0;
+    };
 
-    void setPreferredFramesPerSecond(IPC::Connection::UniqueID, DisplayLinkObserverID, WebCore::FramesPerSecond);
+    DisplayLink(WebCore::PlatformDisplayID);
+    ~DisplayLink();
 
     WebCore::PlatformDisplayID displayID() const { return m_displayID; }
-    
     WebCore::FramesPerSecond nominalFramesPerSecond() const { return m_displayNominalFramesPerSecond; }
 
-    // When responsiveness is critical, we send the IPC to a background queue. Otherwise, we send it to the
-    // main thread to avoid unnecessary thread hopping and save power.
-    static void setShouldSendIPCOnBackgroundQueue(bool value) { shouldSendIPCOnBackgroundQueue = value; }
+    void addClient(Client&);
+    void removeClient(Client&);
+
+    void addObserver(Client&, DisplayLinkObserverID, WebCore::FramesPerSecond);
+    void removeObserver(Client&, DisplayLinkObserverID);
+
+    // FIXME: Maybe callers should just register a DisplayLinkObserverID with the appropriate fps.
+    void incrementFullSpeedRequestClientCount(Client&);
+    void decrementFullSpeedRequestClientCount(Client&);
+
+    void setObserverPreferredFramesPerSecond(Client&, DisplayLinkObserverID, WebCore::FramesPerSecond);
 
 private:
     static CVReturn displayLinkCallback(CVDisplayLinkRef, const CVTimeStamp*, const CVTimeStamp*, CVOptionFlags, CVOptionFlags*, void* data);
     void notifyObserversDisplayWasRefreshed();
 
-    void removeInfoForConnectionIfPossible(IPC::Connection::UniqueID) WTF_REQUIRES_LOCK(m_observersLock);
+    void removeInfoForClientIfPossible(Client&) WTF_REQUIRES_LOCK(m_observersLock);
 
     static WebCore::FramesPerSecond nominalFramesPerSecondFromDisplayLink(CVDisplayLinkRef);
 
     struct ObserverInfo {
         DisplayLinkObserverID observerID;
         WebCore::FramesPerSecond preferredFramesPerSecond;
     };
-    
-    struct ConnectionClientInfo {
+
+    struct ClientInfo {
         unsigned fullSpeedUpdatesClientCount { 0 };
         Vector<ObserverInfo> observers;
     };
 
     CVDisplayLinkRef m_displayLink { nullptr };
     Lock m_observersLock;
-    HashMap<IPC::Connection::UniqueID, ConnectionClientInfo> m_observers WTF_GUARDED_BY_LOCK(m_observersLock);
-    WebCore::PlatformDisplayID m_displayID;
+    HashMap<Client*, ClientInfo> m_observers WTF_GUARDED_BY_LOCK(m_observersLock);
```

Comment:
Why is called `m_observers` although it maps from a `Client` to a `ClientInfo`? Should not it be called `m_clients`?

Why do we use a raw pointer for the key? Should we use WeakPtr<> instead? How do we guarantee the clients will remove themselves before they are destroyed.

Reply:
Done.